serde/src/main/java/com/amazon/ionhiveserde/objectinspectors/utils/IonPrimitiveReader.java

```diff
@@ -0,0 +1,120 @@
+    public static byte byteValue(final String ionPrimitive) {
+        return (byte) buildSingleReader(ionPrimitive, CastType.NUMERIC).longValue();
+    }
+
+    public static BigDecimal decimalValue(final String ionPrimitive) {
+        return buildSingleReader(ionPrimitive, CastType.NUMERIC).bigDecimalValue();
+    }
+
+    public static double doubleValue(final String ionPrimitive) {
+        return buildSingleReader(ionPrimitive, CastType.NUMERIC).doubleValue();
+    }
+
+    public static float floatValue(final String ionPrimitive) {
+        return (float) buildSingleReader(ionPrimitive, CastType.NUMERIC).doubleValue();
+    }
+
+    public static int intValue(final String ionPrimitive) {
+        return (int) buildSingleReader(ionPrimitive, CastType.NUMERIC).longValue();
+    }
+
+    public static long longValue(final String ionPrimitive) {
+        return buildSingleReader(ionPrimitive, CastType.NUMERIC).longValue();
+    }
+
+    public static short shortValue(final String ionPrimitive) {
+        return (short) buildSingleReader(ionPrimitive, CastType.NUMERIC).longValue();
+    }
+
+    // Restricted to temporal ion types
+    public static Date dateValue(final String ionPrimitive) {
+        return buildSingleReader(ionPrimitive, CastType.TEMPORAL).dateValue();
+    }
+
+    public static Timestamp timestampValue(final String ionPrimitive) {
+        Timestamp ts = buildSingleReader(ionPrimitive, CastType.TEMPORAL).timestampValue();
+        return ts;
+    }
+
+    // No restrictions on ion type
+    public static String stringValue(final String ionPrimitive) {
+        return buildSingleReader(ionPrimitive, CastType.ALL).stringValue();
+    }
+
+
+
+    private static IonReader buildSingleReader(final String ionPrimitive, final CastType targetType) {
+        IonReader reader = SYSTEM.newReader(ionPrimitive);
+        IonType type = reader.next();
+        if (!targetType.canCast(type)) {
+            throw new IllegalArgumentException(
+                    String.format("Unable to cast value %s of type %s as %s type.",
+                            ionPrimitive,
+                            type.name(),
+                            targetType.toString().toLowerCase()));
+        }
+        return reader;
+    }
+
+    private enum CastType {
+        ALL,
+        BOOLEAN,
+        NUMERIC,
+        TEMPORAL;
+
+        private static Set<IonType> NUMERIC_TYPE = ImmutableSet.of(IonType.INT, IonType.DECIMAL, IonType.FLOAT);
+
+        public boolean canCast(final IonType type) {
+            switch (this) {
+                case ALL:
+                    return true;
+                case BOOLEAN:
+                    return type == IonType.BOOL;
+                case TEMPORAL:
+                    return type == IonType.TIMESTAMP;
+                case NUMERIC:
+                    return NUMERIC_TYPE.contains(type);
+                default:
+                    return false;
+            }
+        }
+    }
```

Comment:
The JavaDocs on [IonReader](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/IonReader.java) `*value()` methods defines which `IonType`s are compatible with those methods.

For example, `stringValue()` is [only valid](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/IonReader.java#L382) when the reader is positioned on a value with `IonType` `STRING` or `SYMBOL`. Are we trying to be as permissive as possible here? If so, then we may want to add logic to handle conversions that the `IonReader` doesn't handle automatically, but are possible, such as integer-to-string (similar to the logic that was added in https://github.com/amzn/ion-hive-serde/pull/79).

Reply:
Updated to properly cast non string IonValues as strings.